main/src/library/String.flix

```diff
@@ -99,6 +100,1199 @@ namespace String {
+    /// three characters are replaced with ellipses.
+    ///
+    /// If the length of `s` exceeds `w` and `w < 3" then the empty string is returned.
+    ///
+    pub def abbreviateLeft(w: Int32, s: String): String =
+        match length(s) {
+            case len if len < w => s
+            case _   if w < 3   => ""
+            case len            => {
+                let start = len - w + 3;
+                "..." + slice(s, start, len)
+            }
+        }
+
+    ///
+    /// Abbreviate the string `s` if it exceeds the width `w`.
+    ///
+    /// If the length of `s` exceeds `w` and `w >= 3" then `s` is truncated and the last
+    /// three characters are replaced with ellipses.
+    ///
+    /// If the length of `s` exceeds `w` and `w < 3" then the empty string is returned.
+    ///
+    pub def abbreviateRight(w: Int32, s: String): String =
+        match length(s) {
+            case len if len < w => s
+            case _   if w < 3   => ""
+            case _              => slice(s, 0, w - 3) + "..."
+        }
+
+    ///
+    /// Calculate the Levenshtein distance between the strings `s` and `t`.
+    ///
+    /// The answer is the number deletions, insertions or substitutions needed to turn
+    /// string `s` into string `t`.
+    ///
+    pub def levenshteinDistance(s: String, t: String) : Int32 = levenshteinDistanceHelper(s, t) as & Pure
+
+    ///
+    /// Helper function for `levenshteinDistance`.
+    ///
+    /// This is the Wikipedia two-matrix rows algorithm:
+    /// https://en.wikipedia.org/wiki/Levenshtein_distance#Iterative_with_two_matrix_rows
+    ///
+    def levenshteinDistanceHelper(s: String, t: String) : Int32 & Impure =
+        let m = length(s);
+        let n = length(t);
+        let v0 = Array.new(0, n+1);
+        let v1 = Array.new(0, n+1);
+        forIndex(0, n, i -> v0[i] = i );
+        forIndex(0, m-1, i -> {
+            v1[0] = i + 1;
+            forIndex(0, n-1, j -> {
+                let substitutionCost = if (charAt(i, s) == charAt(j, t)) 0 else 1;
+                let a = v1[j] + 1;
+                let b = v0[j + 1] + 1;
+                let c = v0[j] + substitutionCost;
+                v1[j+1] = Int32.min(a, Int32.min(b, c))
+            });
+            arraySwap(v0,v1)
+        });
+        v0[n]
+
+    ///
+    /// Helper function for `levenshteinDistance`.
+    ///
+    /// This implements a forloop, the end point `e` is inclusive.
+    ///
+    def forIndex(b: Int32, e: Int32, f: Int32 ~> Unit): Unit & Impure =
+        if (b <= e) { f(b); forIndex(b+1, e, f) } else ()
+
+    ///
+    /// Helper function for `levenshteinDistance`.
+    ///
+    def arraySwap(a: Array[Int32], b : Array[Int32]): Unit & Impure =
+        forIndex(0, Array.length(a) - 1, {i ->  a[i] = b[i]})
+
+
+    ///
+    /// Test equality of strings `s1` and `s2` ignoring case.
+    ///
+    pub def equalsCI(s1: String, s2: String): Bool =
```

Comment:
Is this meant to be public? If it is then I think we should consider a better name. Possibly we could have an enum for comparing strings with cases like `IgnoreCase`, `IgnoreCaseAndWhiteSpace` or some such. Just brainstorming.